### PR TITLE
feat: add target parameter, make it major by default

### PR DESCRIPTION
**Pull request by fahric**
Closes fastify#44

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/checkTargetMatchToPR.test.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    t.equal(checkTargetMatchToPR(getMajorPRTitle()),true,'When PR is major and no target is provided should return true')
+
+    t.equal(checkTargetMatchToPR(getPatchPRTitle(),'major'),true,'When PR is patch and target is major should return true')
+    t.equal(checkTargetMatchToPR(getMinorPRTitle(),'major'),true,'When PR is minor and target is major  should return true')
+    t.equal(checkTargetMatchToPR(getMajorPRTitle(),'major'),true,'When PR is major and target is major should return true')
+
+    t.equal(checkTargetMatchToPR(getPatchPRTitle(),'minor'),true,'When PR is patch and target is minor should return true')
+    t.equal(checkTargetMatchToPR(getMinorPRTitle(),'minor'),true,'When PR is minor and target is minor should return true')
+    t.equal(checkTargetMatchToPR(getPatchPRTitle(),'patch'),true,'When PR is patch and target is patch should return true')
+    t.end()
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:25]: too many assertions for different scenarios in the same test. use different tests instead. also omit the description in the assertion, it's way too verbose. finally, if you're simply checking for true/false you could consider using simply t.ok and t.notOk

**Review by simoneb — CHANGES_REQUESTED** [review]
See inline comments

**Comment by simoneb** [comment]
Although it's not enforced in this repo, let's try to stick to conventional commits when committing to this repo. Also stick to shorter commit messages as per git's guidelines (below 50 chars on the first line of the commit). Please rebase and squash commits to make it a single one which adheres to the above.

**Comment by simoneb** [comment]
Also please modify the workflow file to run linting and testing and make the automerge job depend on it. Inspire from https://github.com/nearform/get-jwks/blob/main/.github/workflows/ci.yml (no need for caching node_modules though). Note the syntax of the npm lint script looks wrong, I think it should be `eslint .`.

**Comment by simoneb** [comment]
the ci workflow should also run on push, not just pull request. can you please make that change too?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,7 +1,22 @@
 name: CI
-on: pull_request
+on: [push, pull_request]
 jobs:
-  approve:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x, 14.x]
```
Comment (simoneb) [inline .github/workflows/ci.yml:8]: we don't need a matrix, let's use LTS (14.x)

File: src/util.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 'use strict'
 
 const core = require('@actions/core')
+const { getTargetInput } = require('./getTargetInput')
```
Comment (simoneb) [inline src/util.js:4]: put this together with the one below

File: src/checkTargetMatchToPR.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+'use strict'
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:1]: let's use regexp and semver package, also let's figure out if semver considers a change like: from 0.1.0 to 0.2.0 as major or minor (due to the 0 as the first number)

File: test/checkTargetMatchToPR.test.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+'use strict'
+const tap = require('tap')
+const checkTargetMatchToPR = require('../src/checkTargetMatchToPR')
+
+tap.test('getTargetInput', async (t) => {
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:5]: fix test name

Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:5]: prettier arrowParens: 'avoid'

File: test/checkTargetMatchToPR.test.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  const getPatchPRTitle = 'chore(deps-dev): bump fastify from 3.18.0 to 3.18.1'
+  const getMinorPRTitle = 'chore(deps-dev): bump fastify from 3.18.0 to 3.19.1'
+  const getMajorPRTitle = 'chore(deps-dev): bump fastify from 3.18.0 to 4.18.1'
+  const getMajorPRTitleWithMultipleFroms =
+    'chore(deps-dev): bump from fastify from 3.18.0 to 4.18.1'
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:10]: remove the leading "get"

File: src/checkTargetMatchToPR.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    return targetOptions.patch
+  }
+  return null
+}
+const previousWordIsFrom = (titleWordsArray, index) =>
+  titleWordsArray[index - 1] === 'from'
+const previousWordIsTo = (titleWordsArray, index) =>
+  titleWordsArray[index - 1] === 'to'
+const checkTargetMatchToPR = (prTitle, target) => {
+  if (!prTitle || target === targetOptions.major) {
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:24]: we don't need to handle falsy title

File: test/checkTargetMatchToPR.test.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+  t.test('should return true when target is major', async (t) => {
+    t.test('and title is null', async (t) => {
+      t.ok(checkTargetMatchToPR(null, 'major'))
+    })
+    t.test('and title is empty', async (t) => {
+      t.ok(checkTargetMatchToPR('', 'major'))
+    })
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:18]: let's not handle these cases

File: test/checkTargetMatchToPR.test.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    t.test('and PR is patch', async (t) => {
+      t.ok(checkTargetMatchToPR(getPatchPRTitle, 'major'))
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:20]: let's always use the targetOptions object instead of hardcoding strings

**Review by simoneb — COMMENTED** [review]
LGTM, see inline comments

File: src/checkTargetMatchToPR.js (hunk 4 above)
Comment (fahric) [reply src/checkTargetMatchToPR.js:1]: using regex and semver pkg now. semver pks considers 0.1.0 to 0.2.0 as minor udpate.. however documantation here(https://semver.org/#spec-item-4) says, Major version zero (0.y.z) is for initial development. Anything MAY change at any time. i think its best if we treat Major version zero updates as major..

File: src/checkTargetMatchToPR.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+'use strict'
+const semverDiff = require('semver/functions/diff')
+
+const { targetOptions } = require('./getTargetInput')
+
+const checkTargetMatchToPR = (prTitle, target) => {
+  if (target === targetOptions.major) {
+    return true
+  }
+  const expression = new RegExp(/from ([^\s]+) to (.+)/g)
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:10]: remove constructor, global specifier is not needed. also move outside this function as we don't need to recreate it each time.

```suggestion
  const expression = /from ([^\s]+) to (.+)$/
```

File: src/checkTargetMatchToPR.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    if (prTarget) {
+      if (prTarget === targetOptions.major) {
+        return false
+      }
+      if (prTarget === targetOptions.minor && target === targetOptions.patch) {
+        return false
+      }
+    }
+  }
+  return true
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:24]: 1. prefer returning early. if match === null -> return true
2. from the docs

> diff(v1, v2): Returns difference between two versions by the release type (major, premajor, minor, preminor, patch, prepatch, or prerelease), or null if the versions are the same.

it doesn't look like we're accounting for all the possible release types here

**Review by simoneb — CHANGES_REQUESTED** [review]

**Comment by simoneb** [comment]
please rebase on main branch and resolve conflicts

File: src/checkTargetMatchToPR.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+const semverDiff = require('semver/functions/diff')
+
+const { targetOptions, semanticVersionOrder } = require('./getTargetInput')
+
+const expression = /from ([^\s]+) to (.+)$/
+
+const checkTargetMatchToPR = (prTitle, target) => {
+  if (target === targetOptions.major) {
+    return true
+  }
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:11]: do we need this special case? doesn't it fall in the logic below?

File: src/checkTargetMatchToPR.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  const match = expression.exec(prTitle)
+
+  if (match === null) {
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:14]: let's not check against null, let's simply test a falsy value for more robustness

File: src/checkTargetMatchToPR.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    return true
+  }
+  const prTarget = semverDiff(match[1], match[2])
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:17]: let's call this variable simply diff

File: src/checkTargetMatchToPR.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  if (prTarget === null) {
+    return true
+  }
+  if (
+    semanticVersionOrder.indexOf(prTarget) >
+    semanticVersionOrder.indexOf(target)
+  ) {
+    return false
+  }
+  return true
```
Comment (simoneb) [inline src/checkTargetMatchToPR.js:27]: this logic can be simplified with a single return statement: !diff || target <= diff

File: test/checkTargetMatchToPR.test.js [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+  const preMinorPRTitle =
+    'chore(deps-dev): bump fastify from 0.0.1-foo to 0.1.0'
+  const minorPRTitle = 'chore(deps-dev): bump fastify from 3.18.0 to 3.19.1'
+  const majorPRTitle = 'chore(deps-dev): bump fastify from 3.18.0 to 4.18.1'
+  const majorPRTitleWithMultipleFroms =
+    'chore(deps-dev): bump from fastify from 3.18.0 to 4.18.1'
+  const preReleaseUpgradePRTitle =
+    'chore(deps-dev): bump fastify from 3.18.0-alpha to 3.18.1-beta'
+  const preReleaseToPathUpgradePRTitle =
+    'chore(deps-dev): bump fastify from 3.18.0-alpha to 3.18.2'
```
Comment (simoneb) [inline test/checkTargetMatchToPR.test.js:23]: move outside test as they're global constants

File: test/getTargetInput.test.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+'use strict'
+const tap = require('tap')
+
+const { targetOptions } = require('../src/getTargetInput')
+const { getTargetInput } = require('../src/getTargetInput')
```
Comment (simoneb) [inline test/getTargetInput.test.js:5]: use single require

File: src/index.js [hunk 18]
```diff
@@ -36,6 +38,10 @@ async function run() {
       return logWarning('Not a dependabot PR, skipping.')
     }
 
+    const isTargetMatchToPR = checkTargetMatchToPR(pr.title, TARGET)
+    if (!isTargetMatchToPR) {
+      return logWarning('Target specified does not match to PR, skipping.')
+    }
```
Comment (simoneb) [inline src/index.js:44]: add some blank lines for readability and consistency

File: test/getTargetInput.test.js (hunk 17 above)
Comment (fahric) [reply test/getTargetInput.test.js:5]: webstorm being clever here

**Review by simoneb — APPROVED** [review]

File: .taprc [hunk 19]
```diff
@@ -0,0 +1 @@
+coverage: false
```
Comment (climba03003) [inline .taprc:1]: Is it better to turn on coverage and disable check-coverage only? The other three is added for consistency through all repository.

```suggestion
ts: false
jsx: false
flow: false
check-coverage: false
coverage: true
```

**Review by climba03003 — APPROVED** [review]
LGTM.

File: .github/workflows/ci.yml [hunk 20]
```diff
@@ -1,7 +1,19 @@
 name: CI
-on: pull_request
+on: [push, pull_request]
 jobs:
-  approve:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@2.3.4
+      - uses: actions/setup-node@2.1.5
```
Comment (simoneb) [inline .github/workflows/ci.yml:8]: ci fails because these version numbers should be preceded by v, e.g. v2.3.4

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -1,7 +1,19 @@
...
 jobs:
-  approve:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.4
+      - uses: actions/setup-node@v2.1.5
+        with:
+          node-version: '14.x'
+      - run: npm i
```
Comment (simoneb) [inline .github/workflows/ci.yml:11]: let's prefer npm ci when there's a lockfile

```suggestion
      - run: npm ci
```

**Review by simoneb — APPROVED** [review]

**Pull request merged**